### PR TITLE
Attachments and configurable blobs

### PR DESCRIPTION
**Pull request by dimitri-yatsenko**
* Nans are interpreted as NULL for floating-point attributes at insert #467
* implement configurable subfolders in external storage #475 
* implement file attachment datatype #480 
* implement configurable blobs #497 
* add `dj.get_schema_names()`

### CODE REVIEW

File: datajoint/__init__.py [hunk 1]
```diff
@@ -15,29 +15,27 @@
 """
 
 __author__ = "Dimitri Yatsenko, Edgar Y. Walker, and Fabian Sinz at Baylor College of Medicine"
-__date__ = "Nov 15, 2018"
+__date__ = "January 14, 2018"
```
Comment (austin-hilberg) [inline datajoint/__init__.py]: Do you mean 2019?

File: datajoint/declare.py [hunk 2]
```diff
@@ -291,32 +291,41 @@ def compile_attribute(line, in_key, foreign_key_sql):
...
+    if is_configurable:
         if in_key:
-            raise DataJointError('External attributes cannot be primary in:\n%s' % line)
+            raise DataJointError('Configurable attributes cannot be primary in:\n%s' % line)
+        match['comment'] = ':{type}:{comment}'.format(**match)  # insert configurable type into comment
         store_name = match['type'].split('-')
-        if store_name[0] != 'external':
-            raise DataJointError('External store types must be specified as "external" or "external-<name>"')
+        if store_name[0] not in ('external', 'blob', 'attach'):
+            raise DataJointError('Invalid attribute type in:\n%s' % line)
```
Comment (austin-hilberg) [inline datajoint/declare.py]: The previous error message gave a more explicit indication of what to do to fix the problem. I find the new one less informative.

File: datajoint/erd.py [hunk 3]
```diff
@@ -50,8 +50,8 @@ class ERD:
         """
 
         def __init__(self, *args, **kwargs):
-            warnings.warn('ERD functionality depends on matplotlib and pygraphviz. Please install both of these '
-                          'libraries to enable the ERD feature.')
+            warnings.warn('ERD functionality depends on matplotlib, networkx, and pygraphviz. '
+                          'Please install both of these libraries to enable the ERD feature.')
```
Comment (austin-hilberg) [inline datajoint/erd.py]: There are now three libraries mentioned.

File: tests/test_legacy_external_class.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
+from nose.tools import assert_true, assert_list_equal, raises
+from numpy.testing import assert_almost_equal
+import datajoint as dj
+from . import schema_legacy_external as modu
+
+
+def test_heading():
+    heading = modu.Simple().heading
```
Comment (austin-hilberg) [inline tests/test_legacy_external_class.py]: You removed the `()` in line 8 of `test_external_class.py` above. Should you do so here as well?

File: tests/test_legacy_external_class.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    assert_true('item' in heading)
+    assert_true(heading['item'].is_external)
+
+
+def test_insert_and_fetch():
+    original_list = [1, 3, 8]
+    modu.Simple().insert1(dict(simple=1, item=original_list))
```
Comment (austin-hilberg) [inline tests/test_legacy_external_class.py]: Same as above; should the `()` be removed throughout this method?

File: tests/test_legacy_external_class.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    q = (modu.Simple() & {'simple': 1}).fetch1()
+    assert_list_equal(list(q['item']), original_list)
+    # test with cache
+    dj.config['cache'] = previous_cache
+    q = (modu.Simple() & {'simple': 1}).fetch1()
+    assert_list_equal(list(q['item']), original_list)
+
+
+def test_populate():
+    image = modu.Image()
```
Comment (austin-hilberg) [inline tests/test_legacy_external_class.py]: Can we remove these `()` throughout this and the next two methods?

File: tests/test_legacy_external_class.py (hunk 4 above)
Comment (dimitri-yatsenko) [reply tests/test_legacy_external_class.py]: This does not really matter.  Accessing attributes of classes directly still triggers instantiation of an object, so a more efficient way is to instantiate.  The omission of parentheses was provided for user convenience.

File: tests/test_legacy_external_class.py (hunk 6 above)
Comment (dimitri-yatsenko) [reply tests/test_legacy_external_class.py]: Yes, we could but, when using the same table multiple times, it's more efficient to instantiate once.

Comment (dimitri-yatsenko) [reply tests/test_legacy_external_class.py]: This particular snippet is inside `test_legacy_external_class`, which tests old functionality, so it's a good place to test whether the old syntax still works.

File: tests/test_legacy_external_class.py (hunk 5 above)
Comment (dimitri-yatsenko) [reply tests/test_legacy_external_class.py]: This is inside legacy tests.  We support both syntaxes.

File: datajoint/erd.py (hunk 3 above)
Comment (dimitri-yatsenko) [reply datajoint/erd.py]: fixed

File: datajoint/version.py [hunk 7]
```diff
@@ -1 +1 @@
-__version__ = "0.11.1"
+__version__ = "0.12.0"
```
Comment (eywalker) [inline datajoint/version.py]: Please change this to `"0.12.dev"`

**Pull request merged**